### ISSUE TITLE
Authenticating using an AccessToken

### ISSUE THREAD

**Issue by jdarre** [MSEndpointMgr/IntuneWin32App#114]
Thanks for making such a great module!

Now that GDAP is used for accessing tenants, using the Secure Application Module, authenticating using a AccessToken directly would be very useful, I think. Any thoughts on that forward?

**Comment by NickolajA** [comment]
This feels like news to me, could you expand a bit on what you mean?

**Comment by jdarre** [comment]
Sure :-)
I am thinking something similar to the Mg-module (Connect-MgGraph), where you can authenticate using the param -AccessToken

Having an AccessToken parameter, you can create the accessToken first, and then use that to connect directly, without the need for the Get-MsalToken used in the module. This gives the module more flexible ways to authenticating, f ex using the Get-AzAccessToken, or in my case, using a clientId, secret, tenantId and refreshToken to create an accesstoken. The last one is used when you have a multi-tenant app, and are a partner. I.e, I can connect to all tenants where I have permission (GDAP) without a login-prompt for every tenant I want to update an application for.

I tried quickly by adding two new functions described below, and it looks like it is working. I'll see if I can add this into the Connect-MSIntuneGraph.

Function Connect-MSIntuneGraphAccessToken {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory=$true)]
        [string]
        $AccessToken,
        [Parameter(Mandatory = $true)]
        [string]$TenantID,
        [Parameter(Mandatory = $true)]
        $ExpiresOn
    )

    $Global:AccessToken = $AccessToken
    $Global:AccessTokenTenantID = $TenantID
    Write-Verbose -Message "Successfully retrieved access token"
                
    try {
        # Construct the required authentication header
        $Global:AuthenticationHeader = New-AuthenticationHeaderAccessToken -AccessToken $Global:AccessToken -ExpiresOn $ExpiresOn
        Write-Verbose -Message "Successfully constructed authentication header"

        # Handle return value
        return $Global:AuthenticationHeader
    } catch [System.Exception] {
        Write-Warning -Message "An error occurred while attempting to construct authentication header. Error message: $($PSItem.Exception.Message)"
    }
}


function New-AuthenticationHeaderAccessToken {
  
    param(
        [parameter(Mandatory = $true, HelpMessage = "Pass the AuthenticationResult object returned from Get-AccessToken cmdlet.")]
        [ValidateNotNullOrEmpty()]
        [string]$AccessToken,
        [parameter(Mandatory = $true, HelpMessage = "Datetime when token expires")]
        [ValidateNotNullOrEmpty()]
        $ExpiresOn
    )
    Process {
        # Construct default header parameters
        $AuthenticationHeader = @{
            "Content-Type"  = "application/json"
            "Authorization" = $AccessToken
            "ExpiresOn"     = $ExpiresOn
        }

        # Handle return value
        return $AuthenticationHeader
    }
}

**Comment by rodw-wilkins** [comment]
> 

If I'm understanding correctly, I second this.

Right now I have to authenticate separately in AzureAD in order to retrieve and assign groups to the apps - I'm thinking having an access token that can authorize all together without separate login prompts would be much better.

**Comment by NickolajA** [comment]
Have you tested with creating your own app registration, providing it with the required Graph API permissions to read e.g. groups/users etc from Entra ID, but also giving it the required Intune permissions to create Win32 apps? If you do that, you can authenticate this way:

Connect-MSIntuneGraph -TenantID <XX> -ClientID <XX> [-ClientSecret <XX>]

For the multi-tenant purpose, I can see where that makes sense. Let me think about this a bit to see how I can modify the connect function.

**Comment by jdarre** [comment]
> > 
> 
> If I'm understanding correctly, I second this.
> 
> Right now I have to authenticate separately in AzureAD in order to retrieve and assign groups to the apps - I'm thinking having an access token that can authorize all together without separate login prompts would be much better.

Yes, you can use the same AccessToken in e.g. the Mg-module to all the stuff the application have permissions to do.

**Comment by Djaymam** [comment]
thank you so much @jdarre for pointing this, i'm currently trying to find a solution for this, I'm currently working on an application that will publish Intune application on behalf of the user, and prompting user for login credentials not in the table and temporary token would be great in my case. I hope this will be implemented.

Tank you @NickolajA for this amazing module, i'm having a blast with it.
